core/src/main/scala/chisel3/Data.scala

```diff
@@ -435,27 +435,79 @@ abstract class Data extends HasId with NamedComponent with SourceInfoDoc {
+        case (None, None, None, None, None, None) => s""
+      }
+    }
+  }
+
   // User-friendly representation of the binding as a helper function for toString.
   // Provides a unhelpful fallback for literals, which should have custom rendering per
   // Data-subtype.
   // TODO Is this okay for sample_element? It *shouldn't* be visible to users
-  protected def bindingToString: String = Try(topBindingOpt match {
-    case None => ""
-    case Some(OpBinding(enclosure, _)) => s"(OpResult in ${enclosure.desiredName})"
-    case Some(MemoryPortBinding(enclosure, _)) => s"(MemPort in ${enclosure.desiredName})"
-    case Some(PortBinding(enclosure)) if !enclosure.isClosed => s"(IO in unelaborated ${enclosure.desiredName})"
-    case Some(PortBinding(enclosure)) if enclosure.isClosed =>
-      DataMirror.fullModulePorts(enclosure).find(_._2 eq this) match {
-        case Some((name, _)) => s"(IO $name in ${enclosure.desiredName})"
-        case None => s"(IO (unknown) in ${enclosure.desiredName})"
-      }
-    case Some(RegBinding(enclosure, _)) => s"(Reg in ${enclosure.desiredName})"
-    case Some(WireBinding(enclosure, _)) => s"(Wire in ${enclosure.desiredName})"
-    case Some(DontCareBinding()) => s"(DontCare)"
-    case Some(ElementLitBinding(litArg)) => s"(unhandled literal)"
-    case Some(BundleLitBinding(litMap)) => s"(unhandled bundle literal)"
-    case Some(VecLitBinding(litMap)) => s"(unhandled vec literal)"
-  }).getOrElse("")
+  @deprecated("This was never intended to be visible to user-defined types", "Chisel 3.5.0")
+  protected def bindingToString: String = bindingToStringOpt().getOrElse("")
+
+  private[chisel3] def bindingToStringOpt(topBindingOpt: Option[TopBinding] = this.topBindingOpt) =
+    Try({this.topBindingOpt match {
+      case None => None
+      case Some(OpBinding(_, _)) => Some("OpResult")
+      case Some(MemoryPortBinding(_, _)) => Some("MemPort")
+      case Some(PortBinding(_)) => Some("IO")
+      case Some(RegBinding(_, _)) => Some("Reg")
+      case Some(WireBinding(_, _)) => Some("Wire")
+      case Some(DontCareBinding()) => Some("(DontCare)")
+      case Some(ElementLitBinding(litArg)) => Some("(unhandled literal)")
+      case Some(BundleLitBinding(litMap)) => Some("(unhandled bundle literal)")
+      case Some(VecLitBinding(litMap)) => Some("(unhandled vec literal)")
+    }}).getOrElse(None)
+
+  private[chisel3] def nameStrOpt: Option[String] = {
+    val computedName: Option[String] = _computeName(None, Some(""), ".")
+    computedName match {
+      case Some(".") | Some("") => None
+      case _ => computedName
+    }
+  }
+
+  private[chisel3] def subcomponentStrOpt(enclosure: Option[BaseModule]): Option[String] = enclosure match {
+    case None => None
+    case Some(enc) => this.getOptionRef match {
+      case None => None
+      case Some(Index(imm, value)) => Some(s"${imm.simpleName}[${value.localName}]")
+      case _ => None
+    }
+  }
```

Comment:
```suggestion
  private[chisel3] def subcomponentStrOpt(enclosure: Option[BaseModule]): Option[String] =
    enclosure.flatMap { _ => // enclosure is unused
      this.getOptionRef.collect {
        case Index(imm, value) => s"${imm.simpleName}[${value.localName}]"
      }
    }
```
Any time you match on an `Option` and have `None => None`, you can just `map` (or in this case `flatMap`) and any time you have `_ => None` (which also catches the `None => None` case), you can just `collect`.

The slight slylistic changes aside, I'm a bit confused by this method. How does it work when you're using subfields? I'm very surprised to not see the word `Slot` anywhere in this PR so I don't understand how `Record` and `Bundles` are handled correctly.